### ISSUE TITLE
cli, kv: expose ability to `undrain` a node 

### ISSUE THREAD

**Issue by aayushshah15** [cockroachdb/cockroach#57093]
Once a cockroach node is drained using `./cockroach node drain`, we don't support the ability to undrain it. We might want to stop treating `drain` as a one way command because we might want to just temporarily drain the leases away from a node until it stabilizes. The only way to allow the node to start accepting range leases after it's drained is to kill it and bring it back as a fresh node.

The `draining` status of a node is maintained in-memory on each of its stores and in the node's liveness entry. A potential solution would probably revolve around sanely switching this status back.

cc @sumeerbhola

**Comment by tbg** [comment]
Draining a node is undone by restarting the node (note that draining is mosty unrelated to decommissioning). Isn't just restarting the node good enough?

**Comment by knz** [comment]
First before we have any discussion about this -- can we spell out the **why** of this issue?

I seem to understand from an off-hand remark by @sumeerbhola that there's a want for an RPC that "sheds leases but does not stop the node".

That would be a valuable addition *but it has nothing to do with reverting the effect of a drain*.

The drain RPC does many more things besides draining leases, for example it also waits for SQL clients to disconnect and cancels SQL table leases. It's unclear from the issue phrasing whhether "stopping a drain" whould be just about range leases or all these other things as well.

So instead of phrasing this as "undrain a node" please reword the issue as "find a way to enable use case X"

**Comment by sumeerbhola** [comment]
> First before we have any discussion about this -- can we spell out the why of this issue?
> I seem to understand from an off-hand remark by @sumeerbhola that there's a want for an RPC that "sheds leases but does not stop the node".

My original goal was to move all user-facing latency impacting work off a transiently (moderately) unhealthy node (we expected that given an hour or so, the node would return to health, which was going to be shorter than decommissioning it). Draining/undraining was just the mechanism to achieve this goal under the assumption that not being the leaseholder would remove all read traffic and proposal evaluation. This is insufficient when there are follower reads.

**Comment by knz** [comment]
So what we want here is a way to have "kv catch up on background tasks"

and make this happen by "limiting foreground traffic" -- which, in this context means **KV requests** specifically (not just SQL)

So there are a couple of obstacles really:
1. follower reads which can target any replicas, not just leaseholders
2. distsql planning which may use the busy node
3. regular KV requests sent to leaseholder

Here's the panel of approaches considered so far:
- temporarily transfer leases away: only really helps with point 3, moderately with point 2, not with point 1
- block SQL traffic in load balancer: only helps with point 2
- temporarily decommissioning the node (and block the final step of decommissioning, when the status is switched to decommissioned). This does not really solve the primary problem, because a successful decommission needs the node to catch up on background traffic to start with

I think generally the primary problem here needs a new API in KV, where we mark all the replicas as temporarily unavailable from the perspective of distsender as commanded by SQL and distsql planning, but still available for every other use.

@tbg @irfansharif what would you think about that?

Also I wonder the urgency / importance of this, and whether we should include this in the next release's roadmap.

**Comment by lunevalex** [comment]
It's reasonable to build something like this that can be triggered manually, to get our feet wet but ideally a node would detect automatically that it is limping and go into this mode. One such trigger could be failing liveliness X times in some time period, i.e blipping in and out. Then once things catch up and the node proves it can maintain liveliness it can exit this mode and become a full member of the cluster.

**Comment by tbg** [comment]
Something that has been noted to happen in real-world examples is that limping nodes interact poorly with the cluster's desire to balance load. When a node slows down to the point where it is unable to hold node liveness epochs alive, its lease count will drop. Other nodes will then start transferring leases to it. So we would benefit from a "degraded" liveness state, where a node, if it loses liveness, will request a "degraded" liveness state and will remain there for (say) one minute before, assuming liveness was stable over the preceding minute, upgrading its liveness again. We could then avoid upreplicating or lease-transferring to degraded nodes, avoid routing load balancing to degraded nodes, etc.
This almost sounds like the "Draining" field we have today.

**Comment by knz** [comment]
>  This almost sounds like the "Draining" field we have today.

Almost but not quite. The differences are sufficiently large that conflating them is going to introduce confusion and, if effort is made to actually conflate them in the code, headaches and bugs.

I'd suggest we find a new word (“Degrated”) and keep them side by side until we have clear evidence that the concepts overlap.

**Comment by tbg** [comment]
The concepts definitely overlap, but I wasn't suggesting that they should be shoehorned into one.

**Comment by lunevalex** [comment]
@tbg, I was chatting with @aayushshah15 yesterday and we came to the same conclusion. I started working on a quick prototype to see if that works, the code change looks fairly limited and should be pretty straightforward to make. Once I get a working test, I can push a branch for discussion to see if it's something we want to do.

**Comment by andrewbaptist** [comment]
I wanted to resurface this topic again. I've been in a couple of support cases recently where nodes were under heavy stress and the fact that they were leaseholders meant that they caused cluster-level outages (if they were only followers, this wouldn't have happened). 
I understand the reluctance either to overload the meaning of drain (currently it does several things) by adding an undrain, however it might be worth adding a new command that performs these operations. `cockroach node shed`, `cockroach node unshed`.  This would ONLY touch leases on the node to try and remove them all. The SQL port would still be open and replica transfers to this node could still occur. 

For operations like decommission, it would be useful to call this as part of the operation automatically, and for support cases where a single node is causing problems, it could be used to transiently reduce stress.

**Comment by knz** [comment]
yes "shed" would be a good feature. This is also _exactly_ what we need to add to decommissioning in #41017  so a good investment imho.
